design/tlog-spilling.md.html

```diff
@@ -0,0 +1,590 @@
+surfaced this as important enough to include in the initial version of this
+work.
+
+Peeking now works by issuing a range read to the B-tree from `(tag, peek_begin)`
+to `(tag, infinity)`.  This is why the key contains the last version of the
+batch, rather than the beginning, so that a range read from the peek request's
+version will always return all relevant batches.  For each batched tuple, if
+the version is greater than our peek request's version, then we read the commit
+containing that mutation from disk, extract the relevant mutations, and append
+them to our response.  There is a target size of the response, 150KB by
+default.  As we iterate through the tuples, we sum `mutation_bytes`, which
+already informs us how many bytes of relevant mutations we'll get from a given
+commit.  This allows us to make sure we won't waste disk IOs on reads that will
+end up being discarded as unnecessary.
+
+Popping works similarly to before, but now requires recovering information from
+disk.  Previously, we would maintain a map from version to location in the disk
+queue for every version we hadn't yet spilled.  Once spilling has copied the
+value into the B-tree, knowing where the commit was in the disk queue is
+useless to us, and is removed.  In spill-by-reference, that information is
+still needed to know how to map "pop until version 7" to "pop until byte 87" in
+the disk queue.  Unfortunately, keeping this information in memory would result
+in TLogs slowly consuming more and more memory[^versionmap-memory] as more data
+is spilled.  Instead, we issue a range read of the B-tree from `(tag, pop_version)`
+to `(tag, infinity)` and look at the first commit we find with a version
+greater than our own.  We then use its starting disk queue location as the
+limit of what we could pop the disk queue until for this tag.
+
+[^versionmap-memory]: Pessimistic assumptions would suggest that a TLog spilling 1TB of data would require ~50GB of memory to hold this map, which isn't acceptable.
+
+## Detailed Implementation
+
+The rough outline of concrete changes proposed looks like:
+
+0. Allow a new TLog and old TLog to co-exist and be configurable, upgradeable, and recoverable
+0. Modify spilling in new TLogServer
+0. Modify peeking in new TLogServer
+0. Modify popping in new TLogServer
+0. Spill txsTag specially
+
+### Spilling
+
+In spill-by-reference, spilling is now the act of persisting the index of 
+
+### Peeking
+
+A `TLogPeekRequest` contains a `Tag` and a `Version`, and is a request for all
+commits with the specified tag with a commit version greater than or equal to
+the given version.  The goal is to return a 150KB block of mutations.
+
+**************************************************************************
+*                                                                        *
+* +---------+  Tag    +---------+           Tag    +--------+            *
+* |  Peek   |-------->| Spilled | ...------------->| Memory |            *
```

Comment:
How comes its `spilled` first and then `memory` after. Wouldn't you want to go to `memory` only in the happy case and then only go check the btree when you know for sure there is spilled data? Or is just kind of conceptually indicating that you first check if their is spilled data, and if there is, fetch that and then combine it with whatever is in memory?

Reply:
It turns out that trying to pipe lines around just looks weird, so I've tried to emphasize the paralellness in text instead.